src/cli/generator.ts

```diff
@@ -1,41 +1,632 @@
+    ConstantString,
+} from '../language/partialEvaluation/model.js';
+import {
+    abstractResultsOrEmpty,
+    assigneesOrEmpty,
+    blockLambdaResultsOrEmpty,
+    importedDeclarationsOrEmpty,
+    importsOrEmpty,
+    isRequiredParameter,
+    statementsOrEmpty,
+} from '../language/helpers/nodeProperties.js';
+import {group} from 'radash';
+import {IdManager} from '../language/helpers/idManager.js';
+import {isInStubFile} from '../language/helpers/fileExtensions.js';
 
 /* c8 ignore start */
 export const generateAction = async (fileName: string, opts: GenerateOptions): Promise<void> => {
     const services = createSafeDsServices(NodeFileSystem).SafeDs;
     const module = await extractAstNode<SdsModule>(fileName, services);
-    const generatedFilePath = generatePython(module, fileName, opts.destination);
+    const generatedFilePath = generatePython(services, module, fileName, opts.destination);
     // eslint-disable-next-line no-console
     console.log(chalk.green(`Python code generated successfully: ${generatedFilePath}`));
 };
 /* c8 ignore stop */
 
+const RUNNER_CODEGEN_PACKAGE = 'safeds_runner.codegen';
+const PYTHON_INDENT = '    ';
+
+class GenerationInfoFrame {
+    services: SafeDsServices;
+    blockLambdaManager: IdManager<SdsBlockLambda>;
+    importSet: Map<String, ImportData>;
+
+    constructor(services: SafeDsServices, importSet: Map<String, ImportData> = new Map<String, ImportData>()) {
+        this.services = services;
+        this.blockLambdaManager = new IdManager<SdsBlockLambda>();
+        this.importSet = importSet;
+    }
+
+    addImport(importData: ImportData | undefined) {
+        if (importData) {
+            const hashKey = JSON.stringify(importData);
+            if (!this.importSet.has(hashKey)) {
+                this.importSet.set(hashKey, importData);
+            }
+        }
+    }
+
+    getUniqueLambdaBlockName(lambda: SdsBlockLambda): string {
+        return `__block_lambda_${this.blockLambdaManager.assignId(lambda)}`;
+    }
+
+    getServices(): SafeDsServices {
+        return this.services;
+    }
+}
+
+class ImportData {
+    importPath: string;
+    declarationName: string | undefined;
+    alias: string | undefined;
+
+    constructor(
+        importPath: string,
+        declarationName: string | undefined = undefined,
+        alias: string | undefined = undefined,
+    ) {
+        this.importPath = importPath;
+        this.declarationName = declarationName;
+        this.alias = alias;
+    }
+}
+
 export type GenerateOptions = {
     destination?: string;
 };
 
+const getPythonNameOrDefault = function (
+    services: SafeDsServices,
+    object: SdsPipeline | SdsSegment | SdsParameter | SdsDeclaration,
+) {
+    return services.builtins.Annotations.getPythonName(object) || object.name;
+};
+
+const generateAssignee = function (assignee: SdsAssignee): string {
+    if (isSdsBlockLambdaResult(assignee)) {
+        return (assignee as SdsBlockLambdaResult).name;
```

Comment:
The explicit cast is not needed. Since the `isSdsXY` functions [return a type predicate](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates), TypeScript can figure out that `assignee` is an `SdsBlockLambdaResult` inside this branch.